### ISSUE TITLE
Compiler doesn't check for collisions in trait methods

### ISSUE THREAD

**Issue by EleanorNB** [rust-lang/rust#67515]
(This is my first time reporting an issue, so sorry if I get something wrong. This is just something I didn't expect to see.)

The following program compiles and runs without errors or warnings:
``` rs
fn main() {
    problem(34);
}

trait Apple {
    fn twin(&self);
    fn complain(&self);
}

trait Orange {
    fn twin(&self);
    fn report(&self);
}

impl Apple for i32 {
    fn twin(&self) {
        print!("{}", self);
    }
    fn complain(&self) {
        println!("This really shouldn't work");
    }
}

impl Orange for i32 {
    fn twin(&self) {
        print!("Not {}", self);
    }
    fn report(&self) {
        println!("This doesn't seem right");
    }
}

fn problem<T: Apple + Orange>(clash: T) {
    clash.complain();
    clash.report();
}
```

The trait bound in `problem()` leads to a collision of the two `twin()`s, but this isn't detected unless I actually try to call one of them, even when methods from both traits are used. I don't know if this behaviour is intentional or not, but I couldn't find any reference to it anywhere, and it surprised me considering how strict `rustc` usually is.

'rustc --version --verbose`:
rustc 1.40.0 (73528e339 2019-12-16)
binary: rustc
commit-hash: 73528e339aae0f17a15ffa49a8ac608f50c6cf14
commit-date: 2019-12-16
host: x86_64-apple-darwin
release: 1.40.0
LLVM version: 9.0

**Comment by jonas-schievink** [comment]
This is expected behavior. Ambiguous method calls can be disambiguated using the fully-qualified `<Type as Trait>::method()` syntax.

**Comment by mocsy** [comment]
This is ambiguity rather than collision, you need to disambiguate which one you'd like to call.
Here you disambiguate the two implementations, because there is no method overloading in rust (or you could say this is it).
```
impl Apple for i32 {
    fn twin(&self) {
...
    }
}
impl Orange for i32 {
    fn twin(&self) {
...
    }
}
```

Usually having the Trait you want to use is enough to disambiguate the use.
Consider this error in one of my tests:
```
error[E0599]: no method named `call` found for type `impl actix_service::Service` in the current scope
   --> tests/integration_test.rs:443:46
    |
443 |         let response = test::block_fn(|| app.call(request)).unwrap();
    |                                              ^^^^
    |
    = help: items from traits can only be used if the trait is in scope
    = note: the following trait is implemented but not in scope, perhaps add a `use` for it:
            `use actix_service::Service;`
```

In case both of them are in scope, you can use what @jonas-schievink Just wrote while I was typing this.
